File: vaden-frontend/src/state.js
// SPDX-License-Identifier: AGPL-3.0
/*
   Vaden: Versioned Application Deployment Engine
   Copyright (C) 2023 Łukasz Wojtów

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU Affero General Public License as
   published by the Free Software Foundation, either version 3 of the
   License.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU Affero General Public License for more details.

   You should have received a copy of the GNU Affero General Public License
   along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

import { reactive } from 'vue';

/**
 * This is application state. Implementation provides the same functionality as vuex or redux but with slightly
 * better api. Components are not supposed to have own data but use getters instead.
 * Also, state should not be accessed directly from components. Instead, modification functions should be called.
 * Each component has its own 'substate' to keep things organized.
 * See examples in components for usage.
 *
 */
const state = reactive({
    /** Analytics page state
     *
     */
    analytics: {
        dateStart: {},
        dateEnd: {},
        totalVisits: 0,
        stats: [],
    },
    /**
     * /versions page's state
     */
    versions: {
        /**
         * Indicates whether there were some changes made. Used to enable "Update" button
         */
        changed: false,
        /**
         * List of versions. Single version is {'name':name, 'strategy':strategy}
         */
        list: [],
    }
});

export function getAnalytics() {
    return state.analytics
}

export function setAnalyticsResults(analytics) {
    state.analytics.totalVisits = analytics.totalVisits
    state.analytics.stats = analytics.stats
}

export function updateAnalyticsQuery(dateStart, dateEnd) {
    state.analytics.dateStart = dateStart
    state.analytics.dateEnd = dateEnd
}

export function addVersion(name, value) {
    state.versions.list.push({name: name, strategy: value})
}

export function updateVersion(name, value) {
    for (const version of state.versions.list) {
        if (version.name === name) {
            version.strategy = value
            break
        }
    }
    state.versions.changed = true
}

export function listVersions() {
    return state.versions
}

export function deleteVersions() {
    state.versions.changed = false
    state.versions.list.length = 0
}

File: vaden-backend/src/dao/state.rs
// SPDX-License-Identifier: AGPL-3.0
/*
   Vaden: Versioned Application Deployment Engine
   Copyright (C) 2023 Łukasz Wojtów

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU Affero General Public License as
   published by the Free Software Foundation, either version 3 of the
   License.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU Affero General Public License for more details.

   You should have received a copy of the GNU Affero General Public License
   along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

use crate::dao::requests::DailySeed;
use crate::dao::{get_sqlite_connection, Dao};
use crate::error::{Result, VadenError};
use log::{debug, error, info};
use serde::{Deserialize, Serialize};
use sqlx::sqlite::SqliteRow;
use sqlx::{Connection, Row};

pub(crate) trait State: Serialize + for<'a> Deserialize<'a> {
    fn name() -> &'static str;
}

impl Dao {
    /// Gets state from DB. Returns `Ok(None)` if no such value.
    /// Returns error when unable to read DB or unable to deserialize.
    pub(crate) async fn get_state<T: State>(&self, name: &str) -> Result<Option<T>> {
        let row = self.read_row(name).await?;
        match row {
            None => Ok(None),
            Some(row) => deserialize_row(row),
        }
    }

    /// Writes state to DB. In one transaction:
    /// - removes value from DB
    /// - inserts new value (serialized)
    /// Return Err when unable to talk to DB or serialize value.
    pub(crate) async fn set_state<T: State>(&self, state: &T) -> Result<()> {
        let name = T::name();
        debug!("Saving state for {}", name);
        let serialized = serde_json::to_string(state).map_err(|e| {
            error!("Unable to serialize state {}: {}", name, e);
            VadenError::SerializationError
        })?;
        let q = sqlx::query("DELETE FROM state WHERE name = $1").bind(name);
        let url = self.inner.get_read_write_url().await;
        let mut conn = get_sqlite_connection(&url).await?;
        let mut transaction = conn.begin().await.map_err(|e| VadenError::DatabaseError(e.to_string()))?;
        q.execute(&mut *transaction).await.map_err(|e| VadenError::DatabaseError(e.to_string()))?;
        let q = sqlx::query("INSERT INTO state (name, value) VALUES ($1, $2)").bind(name).bind(serialized);
        q.execute(&mut *transaction).await.map_err(|e| VadenError::DatabaseError(e.to_string()))?;
        transaction.commit().await.map_err(|e| VadenError::DatabaseError(e.to_string()))?;
        info!("Saved state for {}", name);
        Ok(())
    }

    async fn read_row(&self, name: &str) -> Result<Option<SqliteRow>> {
        let url = self.inner.get_read_only_url().await;
        let mut connection = get_sqlite_connection(&url).await?;
        let state = sqlx::query("SELECT value FROM state WHERE name = $1")
            .bind(name)
            .fetch_optional(&mut connection)
            .await
            .map_err(|e| VadenError::DatabaseError(e.to_string()))?;
        Ok(state)
    }
}

fn deserialize_row<T: State>(row: SqliteRow) -> Result<Option<T>> {
    let value = row.try_get::<String, _>("value").map_err(|_| VadenError::DatabaseError("No column 'value' in `state` table".into()))?;
    match serde_json::from_str(&value) {
        Ok(v) => Ok(Some(v)),
        Err(e) => {
            error!("Unable to deserialize {} from value in DB: `{}`: {}", T::name(), value, e);
            Err(VadenError::SerializationError)
        }
    }
}

impl State for DailySeed {
    fn name() -> &'static str {
        "DailySeed"
    }
}
